sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java

```diff
@@ -0,0 +1,186 @@
+      throws IOException {
+    // Connection- or Channel-level, depending
+    Object cause = e.getReference();
+    if (cause instanceof Channel) {
+      closeChannel(lesseeId);
+    } else {
+      close();
+    }
+  }
+
+  @Override
+  public void closeChannel(UUID lessee) {
+    Channel toClose = channelsByLessee.remove(lessee);
+    if (toClose != null) {
+      try {
+        toClose.close();
+      } catch (IOException | TimeoutException e) {
+        // ignore
+      }
+    }
+  }
+
+  private Channel getChannel(UUID lessee) throws IOException {
+    if (connection == null) {
+      synchronized (this) {
+        if (connection == null) {
+          ConnectionFactory connectionFactory = new ConnectionFactory();
+          try {
+            connectionFactory.setUri(uri);
+            connectionFactory.setAutomaticRecoveryEnabled(true);
+            connectionFactory.setConnectionTimeout(60000);
+            connectionFactory.setNetworkRecoveryInterval(5000);
+            connectionFactory.setRequestedHeartbeat(60);
+            connectionFactory.setTopologyRecoveryEnabled(true);
+            connectionFactory.setRequestedChannelMax(0);
+            connectionFactory.setRequestedFrameMax(0);
+          } catch (URISyntaxException e) {
+            // full URI excluded lest it contain user/pass
+            throw new IOException("Unable to connect to rabbit; invalid URI (uri redacted)", e);
+          } catch (NoSuchAlgorithmException | KeyManagementException e) {
+            throw new IOException(
+                "Security issue while connecting to rabbit: " + e.getMessage(), e);
+          }
+
+          try {
+            connection = connectionFactory.newConnection();
+          } catch (TimeoutException e) {
+            throw new IOException("Timed out attempting to connect to rabbit", e);
+          }
+        }
+      }
+    }
+
+    return channelsByLessee.computeIfAbsent(
+        lessee,
+        uuid -> {
+          try {
+            return connection
+                .openChannel()
+                .orElseThrow(() -> new RuntimeException("No RabbitMQ channel available"));
+          } catch (IOException e) {
+            throw new RuntimeException("No RabbitMQ channel available", e);
+          }
+        });
+  }
+
+  /**
+   * Closes all Channels, closes the connection, and clears the mapping of Channels by lessee.
+   *
+   * @throws IOException if an error occurs closing the underlying Connection. exceptions thrown
+   *     while closing individual Channels are ignored
+   */
+  @Override
+  public synchronized void close() throws IOException {
+    channelsByLessee.forEach(
+        (id, channel) -> {
+          try {
+            channel.close();
+          } catch (Exception e) {
+            /* ignore */
+          }
+        });
+    channelsByLessee.clear();
+
+    if (connection != null) {
+      try {
+        connection.close();
```

Comment:
It would implicitly do so, but not very elegantly. It's going to be better upstream for the RabbitMq server, to be told explicitly that a Channel is being shut down so it can free resources and mark messages as available for immediate redelivery. It's probably not particularly important, but it should be a slight improvement over purely closing the Connection, I think.